profiles/events.py

```diff
@@ -0,0 +1,27 @@
+import logging
+from typing import Callable
+
+from elife_bus_sdk.events import ProfileEvent
+from elife_bus_sdk.publishers import EventPublisher
+from flask import Flask
+
+from profiles.exceptions import UpdateEventFailure
+from profiles.models import Profile
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+def send_update_events(publisher: EventPublisher) -> Callable[..., None]:
+    def wrapper(sender: Flask, changes: tuple) -> None:  # pylint:disable=unused-argument
+        for instance, operation in changes:
+            if isinstance(instance, Profile):
```

Comment:
Does this cater for something like just an `Affiliation` change? We could do with tests here.

Reply:
Currently I only act if the instance being saved is of type `Profile` though all model changes will pass through here so can be handled. Shall I add support for `Affiliation` and some tests surrounding it?

Reply:
It needs to be if _any_ part of a `Profile` changes (so `EmailAddress` too; what happens with composite types?)

Reply:
OK so `Profile`, `Affiliation` and `EmailAddress` will need to have slightly different handling but the composite type updates effectively count as field updates on the parent model they are attached to. So an `Address` thats changed and saved will trigger the associated `Affiliation` to be updated. I will add them and run the additional tests.

Reply:
OK so we now have support for updates of all models excluding `OrcidToken`.

I had refactor the event publisher function to keep a list of ids for a single `session.commit` and then filter out duplicates so you don't receive multiple messages for single events, for example:

- Updating a `Profile` directly would send one change event` [(<Profile test>, 'update')]` this is fine, one event is sent to SNS.

- Creating and adding an `Affiliation` would send two change events `[(<Affiliation 1>, 'insert'), (<Profile test>, 'update')]` in this instance only one event is sent to SNS because I filter out duplicate related profile ids.

- Updating an `Address` (or any composite types) would only send one change event (linked to the parent model) `[(<Affiliation 1>), 'update']` this is fine and one event will be sent to SNS.

In short, any sort of change or collection of changes to a single `Profile` will send only one event to SNS (including composite types).